### PR TITLE
Use webpack4 hooks

### PR DESCRIPTION
**Pull request by Industrial**

### CODE REVIEW

**Comment by Industrial** [comment]
Looks like this project would need an update to babel7 :)

**Comment by Industrial** [comment]
I didn't update to babel7, but the same test is failing now.

**Comment by Industrial** [comment]
Hmm, that's weird: https://travis-ci.org/60frames/webpack-hot-server-middleware/jobs/348935426#L520

I'm not sure what to do now.

File: src/index.js [hunk 1]
```diff
@@ -128,7 +128,7 @@ function webpackHotServerMiddleware(multiCompiler, options) {
     let serverRenderer;
     let error = false;
 
-    multiCompiler.plugin('done', multiStats => {
+    multiCompiler.hooks.done.tap('WebpackHotServerMiddleware', multiStats => {
```
Comment (richardscarrott) [inline src/index.js]: I'd like to support both APIs as I don't think it'd be much effort, something like this:

```js
const doneHandler = (multiStats) => { ... }
if (multiCompiler.hooks) {
    multiCompiler.cooks.done.tap('WebpackHotServerMiddleware', doneHandler);
} else {
    multiCompiler.plugin('done', doneHandler);
}
```

I'd also like to test with at least both versions of the API, perhaps webpack 3 and webpack 4 -- I know yarn supports aliases out of the box allowing you to require two versions of the same package... not sure if npm supports this in though?

File: test/fixtures/esmodules/dist/client.js [hunk 2]
```diff
@@ -0,0 +1,85 @@
+/******/ (function(modules) { // webpackBootstrap
```
Comment (richardscarrott) [inline test/fixtures/esmodules/dist/client.js]: This can be deleted.

**Comment by richardscarrott** [comment]
Also, I'll take a look at the CI error shortly.

File: src/index.js (hunk 1 above)
Comment (Industrial) [reply src/index.js]: Supporting both versions transparently is a good idea. I've implemented the change.

I'm not sure on the package.json / npm side. I'm checking that now.

File: test/fixtures/esmodules/dist/client.js (hunk 2 above)
Comment (Industrial) [reply test/fixtures/esmodules/dist/client.js]: Done.

**Comment by Industrial** [comment]
I have found the package `test-all-versions` which allowed me to set up tests for webpack 3 and 4.

Right now, the webpack config is different for 3 and 4. How would you like to structure the tests for webpack 3 and 4?

**Comment by richardscarrott** [comment]
@Industrial thanks for your efforts!

I feel `test-all-versions` is a little too involved / requires quite a lot of buy-in -- I think I'd rather migrate to `yarn` and use their aliases so we could do something like this:

``` js
// package.json
{
  "devDependencies": {
    "webpack": "^4.0.0",
    "webpack-legacy": "npm:webpack@^3.0.0",
  }
}

// index.test.js
const webpack = require('webpack');
const webpackLegacy = require('webpack-legacy');
...
```

re: the webpack config between the two versions being incompatible, I think it'd be fine to just create a new fixture called 'webpacklegacy' which is the same as the old esmodule fixture.

If this is too much effort up front we can go ahead with what you've got and I'll create a separate ticket to migrate to yarn?

**Comment by Industrial** [comment]
Glad to help :+1: 

`test-all-versions` was just something I found, so I'll undo it and move to yarn. I'll create the new fixture for the old webpack version next to the existing ones too.

**Comment by Industrial** [comment]
@richardscarrott not sure why it fails.

Also, if the package.json now references 4.0, how will someone actually use it with 3?

**Comment by Industrial** [comment]
Now the `package.json` is correct for webpack, @richardscarrott, I hope you could try this test =)

**Comment by richardscarrott** [comment]
@Industrial I've sent a PR into your branch to fix up the unit tests, at which point this should be good to go -- https://github.com/Industrial/webpack-hot-server-middleware/pull/1

**Comment by Industrial** [comment]
Done

**Comment by richardscarrott** [comment]
@Industrial awesome, thanks for the PR -- it's now available at `webpack-hot-server-middleware@0.5.0` 🎉

**Pull request merged**